File: src/Core/Hat.fs
namespace Zeta.Core

/// **`Hat` — a role-scoped bundle of the whole clarity engine (Aaron 2026-06-08, shadow*).**
///
/// Aaron: *"hats have **lenses**, **suggested landmarks for lens parameters**, **action restrictions**,
/// **uncertainty-reduction traversals**, and **control of other hats/agents wearing those hats**."* A hat (a
/// persona/role) is the engine, scoped to a role — every piece we built, packaged:
///   - **`Lenses`** — the role's `LensRouter` views (what this role attends to);
///   - **`Landmarks`** — suggested **solid ground** (`SolidGround`) the hat's lenses are parameterized by (the
///     role's constants/clocks to navigate by);
///   - **`AllowedActions`** — **action restrictions**: the permitted `ActionGrammar` subset (the role's
///     authority/permissions; empty = unrestricted). Ties to standing-authorization / gated-classes governance;
///   - **`Traversals`** — the role's uncertainty-reduction repertoire (`Traversal`s it can run);
///   - **`Controls`** — names of other hats/agents this hat **controls/coordinates** (the architect-hat-orchestrates
///     / delegation pattern; the meta-control edge).
///
/// Hats **compose** (qubit combinations of hats, #7140): a multi-agent fleet's joint state is the tensor of its
/// agents' worn hats; a hat's `Controls` is the coordination edge between them. So the world-state clarity engine,
/// the meta-controller, and the fleet are the same machinery viewed at the role layer.
///
/// **Honest scope (peel):** a hat is a *declarative bundle* — it says what the role attends to / may do / can run
/// / controls; the *policy* (which lens/traversal/action to pick) is `Salience`/`MetaController`/`ControlMerge`
/// under the intrinsic objective. `AllowedActions` is a structural allow-list (empty = unrestricted), not a proof
/// of authority — real authorization stays the human-gated governance model. `Controls` is a name edge, not an
/// enforcement mechanism. Deterministic (DST).
[<RequireQualifiedAccess>]
module Hat =

    /// A hat's scope (Aaron 2026-06-08): most hats are **game-specific** (scoped to one game/task); some **survive
    /// into the meta** — those are **personas**: general, cross-game, can play *all* games. Personas = the
    /// meta-surviving hats; game-specific hats are ephemeral, scoped to their game.
    type Scope =
        | GameSpecific // scoped to one game/task (many hats)
        | Meta // survives into the meta = a persona (plays all games)

    /// A role/persona bundle: lenses + landmarks + action restrictions + traversals + control edges + scope.
    type Hat<'r> =
        { Name: string
          /// Game-specific, or Meta (a persona that survives into the meta and plays all games).
          Scope: Scope
          Lenses: LensRouter.Lens list
          /// Suggested solid-ground landmarks for lens parameters (cell → its ground kind).
          Landmarks: (string * SolidGround.Ground) list
          /// The permitted action subset (action restriction). Empty = unrestricted.
          AllowedActions: bool[] list
          /// The role's uncertainty-reduction traversals.
          Traversals: Traversal.Traversal<'r> list
          /// Names of other hats/agents this hat controls/coordinates.
          Controls: string list }

    /// Is this hat a **persona** — meta-surviving, plays all games?
    let isPersona (hat: Hat<'r>) : bool = hat.Scope = Meta

    /// The personas (meta-surviving hats) among a set.
    let personas (hats: Hat<'r> list) : Hat<'r> list = hats |> List.filter isPersona

    /// The game-specific hats among a set.
    let gameSpecific (hats: Hat<'r> list) : Hat<'r> list =
        hats |> List.filter (fun h -> h.Scope = GameSpecific)

    /// **Where a hat lives (Aaron 2026-06-08, MUMPS scoping):** a normal (game-specific) hat is **tied to its game
    /// fingerprint** — game-scoped, transient, like a MUMPS *local*. A **meta hat lives in the persistent
    /// substrate, outside game scope, in global/meta scope** — a MUMPS *global* (`^`), surviving across games
    /// (anti-ephemerality, manifesto §5). `address gameKey hat` is the hat's location in the persistent
    /// multidim array: a meta hat is a global (game-independent); a game-specific hat is scoped under `gameKey`
    /// (its game fingerprint, `GameFingerprint`).
    let address (gameKey: string) (hat: Hat<'r>) : string =
        match hat.Scope with
        | Meta -> "^hat/" + hat.Name // MUMPS global: persistent, game-independent (meta scope)
        | GameSpecific -> "game/" + gameKey + "/hat/" + hat.Name // local: scoped to the game fingerprint

    /// Does this hat permit `action`? (Empty `AllowedActions` ⇒ unrestricted ⇒ always true.)
    let permits (action: bool[]) (hat: Hat<'r>) : bool =
        List.isEmpty hat.AllowedActions || List.contains action hat.AllowedActions

    /// Filter a candidate action list to those this hat permits (the role's action restriction applied).
    let restrict (actions: bool[] list) (hat: Hat<'r>) : bool[] list =
        if List.isEmpty hat.AllowedActions then actions
        else actions |> List.filter (fun a -> List.contains a hat.AllowedActions)

    /// Does this hat control the named other hat/agent?
    let controls (other: string) (hat: Hat<'r>) : bool = List.contains other hat.Controls

    /// The solid-ground landmark cells this hat suggests as lens parameters (just the names).
    let landmarkCells (hat: Hat<'r>) : string list = hat.Landmarks |> List.map fst

File: src/Core.TypeScript/hat/hat.ts
/**
 * hat.ts — TypeScript oracle for `src/Core/Hat.fs`.
 *
 * F# is canonical; this conforms. Pinned by `hat-treaty-transcript.json` (generated here, replayed
 * against the F# by `tests/Tests.FSharp/HatTreaty.Tests.fs`).
 *
 * ── What a hat is ────────────────────────────────────────────────────────────
 * A hat is a ROLE: a declarative bundle of lenses (what it attends to), landmarks (the solid ground
 * its lenses are parameterised by), an ACTION RESTRICTION, uncertainty-reduction traversals, and
 * control edges to other hats. It is the ATOMIC base — hats do not nest. Personas compose them
 * (`persona.ts`).
 *
 * ── The authority model, and why it is a set of actions ──────────────────────
 * `allowedActions` is a list of 16-bool `Action`s over the action grammar, and EMPTY MEANS
 * UNRESTRICTED — not "permits nothing". That inversion is deliberate in the F# and is copied
 * exactly: an unconfigured hat is a hat with no restriction, so adding the restriction machinery
 * cannot accidentally lock down every existing role. Read it as "the restriction is absent", never
 * as "the permission set is empty".
 *
 * Note this is a structural ALLOW-LIST, not a proof of authority — the F# says so plainly
 * ("a structural allow-list, not a proof of authority — real authorization stays the human-gated
 * governance model"), and `controls` is a NAME EDGE, not an enforcement mechanism. Neither claim is
 * strengthened in the port; a port that quietly promoted an advisory field into an enforced one
 * would be a security claim invented in translation.
 *
 * ── Generic payloads ─────────────────────────────────────────────────────────
 * `Lens`, `Ground` and `Traversal` are separate F# modules that this slice does not port. They are
 * carried as opaque type parameters rather than modelled, so nothing about them is faked: every
 * function here that touches them (`landmarkCells`, and the persona-level unions) only ever moves
 * them around or compares them, exactly as the F# does.
 */

import { containsAction, type Action } from "./action-grammar";

/**
 * A hat's scope. Most hats are game-specific; some survive into the meta — those are available as
 * personas' meta roles. NOTE the F# correction carried over verbatim: `Meta` means the hat is
 * META-AVAILABLE, not that it IS a persona. The persona is the entity that wears it (`persona.ts`).
 */
export type HatScope = "GameSpecific" | "Meta";

/** A landmark: a cell name paired with its solid-ground kind. */
export type Landmark<G> = readonly [string, G];

/** A role/persona bundle: lenses + landmarks + action restrictions + traversals + control edges + scope. */
export interface Hat<L = unknown, G = unknown, T = unknown> {
  readonly name: string;
  /** Game-specific, or Meta (survives into the meta and plays all games). */
  readonly scope: HatScope;
  readonly lenses: readonly L[];
  /** Suggested solid-ground landmarks for lens parameters (cell → its ground kind). */
  readonly landmarks: readonly Landmark<G>[];
  /** The permitted action subset (action restriction). EMPTY = UNRESTRICTED. */
  readonly allowedActions: readonly Action[];
  /** The role's uncertainty-reduction traversals. */
  readonly traversals: readonly T[];
  /** Names of other hats/agents this hat controls/coordinates. A name edge, not enforcement. */
  readonly controls: readonly string[];
}

/** Is this hat a persona-available (meta-surviving) role? */
export function isPersona<L, G, T>(hat: Hat<L, G, T>): boolean {
  return hat.scope === "Meta";
}

/** The meta-surviving hats among a set. */
export function personas<L, G, T>(hats: readonly Hat<L, G, T>[]): readonly Hat<L, G, T>[] {
  return hats.filter(isPersona);
}

/** The game-specific hats among a set. */
export function gameSpecific<L, G, T>(hats: readonly Hat<L, G, T>[]): readonly Hat<L, G, T>[] {
  return hats.filter(h => h.scope === "GameSpecific");
}

/**
 * Where a hat lives (MUMPS scoping): a meta hat is a GLOBAL (`^hat/<name>`) — persistent and
 * game-independent; a game-specific hat is a LOCAL, scoped under its game fingerprint. The string
 * shapes are part of the treaty and are compared byte-for-byte against the F#.
 */
export function address<L, G, T>(gameKey: string, hat: Hat<L, G, T>): string {
  return hat.scope === "Meta" ? `^hat/${hat.name}` : `game/${gameKey}/hat/${hat.name}`;
}

/** Does this hat permit `action`? Empty `allowedActions` ⇒ unrestricted ⇒ always true. */
export function permits<L, G, T>(action: Action, hat: Hat<L, G, T>): boolean {
  return hat.allowedActions.length === 0 || containsAction(hat.allowedActions, action);
}

/** Filter candidate actions to those this hat permits (the role's action restriction applied). */
export function restrict<L, G, T>(actions: readonly Action[], hat: Hat<L, G, T>): readonly Action[] {
  if (hat.allowedActions.length === 0) return actions;
  return actions.filter(a => containsAction(hat.allowedActions, a));
}

/** Does this hat control the named other hat/agent? */
export function controls<L, G, T>(other: string, hat: Hat<L, G, T>): boolean {
  return hat.controls.includes(other);
}

/** The solid-ground landmark cells this hat suggests as lens parameters (just the names). */
export function landmarkCells<L, G, T>(hat: Hat<L, G, T>): readonly string[] {
  return hat.landmarks.map(([cell]) => cell);
}
